src/main/java/karuna/karuna_backend/content/domain/ContentService.java

```diff
@@ -32,8 +40,55 @@ private ContentDTO getContentByPageFromDatabase(String page) {
         return convertToDTO(allContent);
     }
 
+
+    public MassContentWrapper massUpdateContent(MassContentWrapperRequest massContentWrapperRequest) {
+        return massContentOperation(massContentWrapperRequest, this::filterNotExistedInDatabaseAndUpdatedExisted);
+    }
+
+    public MassContentWrapper massAddContent(MassContentWrapperRequest massContentWrapperRequest){
+        return massContentOperation(massContentWrapperRequest, this::addContentIfNotInDb);
+    }
+
+    private MassContentWrapper massContentOperation(MassContentWrapperRequest request, Predicate<MassContentDto> predicate){
+        //TODO: Sanitize inputs for XSS
+        Map<Boolean, List<MassContentDto>> partitionedContent = request.contents().stream()
+                .collect(Collectors.partitioningBy(predicate));
+        List<MassContentDto> validContent = partitionedContent.get(true);
+        List<MassContentDto> invalidContent = partitionedContent.get(false);
+        return new MassContentWrapper(validContent, invalidContent);
+    }
+
+    private boolean addContentIfNotInDb(MassContentDto massContentDto) {
+        //TODO: Consider if batching would not be a better approach
+        return contentRepository.getByPageAndKey(massContentDto.page(), massContentDto.key())
+                .map(content -> false)
+                .orElseGet(() -> {
+                    Content newContent = mapDtoToContent(massContentDto);
+                    contentRepository.save(newContent);
+                    return true;
+                });
+    }
+
+    @Transactional
```

Comment:
The @Transactional annotation is typically used on methods with public visibility. As of 6.0, protected or package-visible methods can also be made transactional for class-based proxies by default. Note that transactional methods in interface-based proxies must always be public and defined in the proxied interface.